analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyNullableValueAccessBase.cs

```diff
@@ -0,0 +1,52 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
+
+public abstract class EmptyNullableValueAccessBase : SymbolicRuleCheck
+{
+    internal const string DiagnosticId = "S3655";
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var operationInstance = context.Operation.Instance;
+        if (operationInstance.Kind == OperationKindEx.PropertyReference
+            && operationInstance.ToPropertyReference() is var reference
+            && reference.Property.Name == nameof(Nullable<int>.Value)
+            && reference.Instance.Type.IsNullableValueType()
+            && context.HasConstraint(reference.Instance, ObjectConstraint.Null))
+        {
+            ReportIssue(reference.Instance, reference.Instance.Syntax.ToString());
+        }
+        else if (operationInstance.Kind == OperationKindEx.Conversion
```

Comment:
Did anyone had a look why we miss the implicit conversion here? Do we track the work on these False negatives somewhere?

Reply:
I only moved the code from the C# implementation to the base class. I'm not aware of the current state of the implementation.
@antonioaversa Do you know?

Reply:
@martin-strecker-sonarsource 
Concerning VB.NET: implicit conversions do occur, but they result into `ConversionOperation` instances at CFG level. False negatives detected by the rule (such as the one reported [here](https://github.com/SonarSource/sonar-dotnet/pull/6976#pullrequestreview-1353050660)) were mainly due to `ShouldExecute` recall issues, which I think have been discussed in a dedicated meeting.

Concerning C#: unlike VB.NET, basic implicit conversions from `T?` to `T` where `T : struct` are not allowed.
```cs
int? x = 42;
int y = x; // Doesn't compile
Method(x); // Doesn't compile either
```

The only implicit conversion I am aware of, is the one based on `foreach`:
```cs
foreach (int z in new int?[] { null }) { }
foreach (int z in new List<int?> { null }) { }
// etc...
```

Those cases have been documented as false negatives [here](https://github.com/SonarSource/sonar-dotnet/blob/feature/SE/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs#L213). 
They seem to me very narrow, and not deserving more attention for the time being. What do you think?
I am not aware of other cases of implicit conversion from `T?` to `T` where `T : struct`.

Pattern matching should be OK, since `is` would just not match, and does not seem to ever call `.Value`.
Other cases you can think of @martin-strecker-sonarsource ?

Reply:
An interesting case, I have found out about, trying to force a call to `.Value` on a `int?` via pattern matching:
![image](https://user-images.githubusercontent.com/2676742/227569018-bcd18dbf-e1d0-4c62-9279-8ed225547f63.png)

Reply:
C# philosophie is to make cast explicit whenever they can fail. The `foreach` case is an interesting find. The spec is especially tolerant for foreach and is pattern-based so you can e.g. iterate over `Span<T>` even though it doesn't implement any interfaces. Also, explicit conversions are allowed for the iterator variable:

> In the preceding form, type T of a collection element must be implicitly or explicitly convertible to type V of an iteration variable. If an explicit conversion from T to V fails at run time, the foreach statement throws an [InvalidCastException](https://learn.microsoft.com/en-us/dotnet/api/system.invalidcastexception). For example, if T is a non-sealed class type, V can be any interface type, even the one that T doesn't implement. At run time, the type of a collection element may be the one that derives from T and actually implements V. If that's not the case, an [InvalidCastException](https://learn.microsoft.com/en-us/dotnet/api/system.invalidcastexception) is thrown.

[Source](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements#the-foreach-statement)

The only other instance I can think of is in query comprehension syntax:
```cs
    IEnumerable<int?> col =  null;
    _ = from int i in col
        select i;
```
But this is translated to [`Enumerable.Cast<TResult>`](https://learn.microsoft.com/de-de/dotnet/api/system.linq.enumerable.cast?view=net-7.0) under the hood so it is quite different.

Reply:
Indeed, `Enumerable.Cast<TResult>` internally calls:
```cs
private static IEnumerable<TResult> CastIterator<TResult>(IEnumerable source)
{
    foreach (object obj in source)
    {
        yield return (TResult)obj;
    }
}
```

`foreach` boxes the `int?` into a heap object if `obj` is a non-empty struct, and converts it into a null reference if `obj` is an empty struct. Then, casting `obj` into `TResult` results into an unboxing which raises a `NullReferenceException`.

```
.method private final hidebysig specialname newslot virtual 
            instance object System.Collections.IEnumerator.get_Current () cil managed 
        {
            ...
            IL_0000: ldarg.0
            IL_0001: ldfld valuetype [System.Runtime]System.Nullable`1<int32> C/'<CastIterator>d__0'::'<>2__current'
            IL_0006: box valuetype [System.Runtime]System.Nullable`1<int32>
            IL_000b: ret
        } // end of method '<CastIterator>d__0'::System.Collections.IEnumerator.get_Current

 .method private final hidebysig newslot virtual 
            instance bool MoveNext () cil managed 
        {
            ...
                IL_0041: ldarg.0
                IL_0042: ldarg.0
                IL_0043: ldfld class [System.Runtime]System.Collections.IEnumerator C/'<CastIterator>d__0'::'<>s__1'
                IL_0048: callvirt instance object [System.Runtime]System.Collections.IEnumerator::get_Current()
                IL_004d: stfld object C/'<CastIterator>d__0'::'<obj>5__2'
                IL_0052: nop
                IL_0053: ldarg.0
                IL_0054: ldarg.0
                IL_0055: ldfld object C/'<CastIterator>d__0'::'<obj>5__2'
                IL_005a: unbox.any valuetype [System.Runtime]System.Nullable`1<int32>
                IL_005f: stfld valuetype [System.Runtime]System.Nullable`1<int32> C/'<CastIterator>d__0'::'<>2__current'
                IL_0064: ldarg.0
                IL_0065: ldc.i4.1
                IL_0066: stfld int32 C/'<CastIterator>d__0'::'<>1__state'
                IL_006b: ldc.i4.1
                IL_006c: stloc.0
                IL_006d: leave.s IL_00a6

               ...
        } // end of method '<CastIterator>d__0'::MoveNext
```